### PR TITLE
Codec pipeline

### PR DESCRIPTION
**Pull request by ivandruzhinin**

### CODE REVIEW

File: src/main/kotlin/com/exactpro/th2/codec/Main.kt [hunk 1]
```diff
@@ -56,46 +56,52 @@ class CodecCommand : CliktCommand() {
...
+            val configuration = Configuration.create(commonFactory, codecSettings)
+            val applicationContext = ApplicationContext.create(configuration, commonFactory).apply { resources += ::close }
+            val messageRouter = commonFactory.messageRouterMessageGroupBatch
+            val eventRouter = commonFactory.eventBatchRouter
 
-            val rootEventId = createAndStoreRootEvent(applicationContext)
+            val rootEventId = eventRouter.storeEvent(
+                Event.start().apply {
+                    name("Codec_${load<IPipelineCodec>()::class.java.simpleName}_${LocalDateTime.now()}")
+                    type("CodecRoot")
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/Main.kt:77]: Maybe print the name of the handled protocol.

File: src/main/kotlin/com/exactpro/th2/codec/api/impl/ThreadSafeCodec.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        synchronized(this) {
+            decode(messageGroup)
+        }
+    }
+
+    private fun getInstance() = instances.computeIfAbsent(Thread.currentThread()) {
+        codecFactory().apply {
+            init(dictionary, settings)
+        }
+    }
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/api/impl/ThreadSafeCodec.kt:53]: Maybe replace to thread-local

**Review by Nikita-Smirnov-Exactpro — DISMISSED** [review]

File: src/main/kotlin/com/exactpro/th2/codec/api/impl/ThreadSafeCodec.kt (hunk 2 above)
Comment (ivandruzhinin) [reply src/main/kotlin/com/exactpro/th2/codec/api/impl/ThreadSafeCodec.kt:53]: Please take a look at `close()` method

File: src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            val parentEventId = messageGroup.parentEventId
+
+            messageGroup.runCatching(codec::decode).onSuccess {
+                if (it.messagesCount < messageGroup.messagesCount) {
+                    parentEventId.onEvent("Decoded message group contains less messages ($it.messagesCount) than encoded one (${messageGroup.messagesCount})")
+                }
+
+                messageBatch.addGroups(it)
+            }.onFailure {
+                parentEventId.onEvent("Failed to decode message group: ${messageGroup.toDebugString()}", it)
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt:44]: 'messageGroup' may be very large. I think we can attach parsed message via message-id

Comment (ivandruzhinin) [reply src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt:44]: But which message-id should we take? There could be many messages in a message group

Comment (ivandruzhinin) [reply src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt:44]: NVM, an event can contain multiple message-ids

Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt:44]: We can attach all of them

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**